src/daft-io/src/multipart.rs

```diff
@@ -0,0 +1,108 @@
+use std::{io::Write, num::NonZeroUsize};
+
+use async_trait::async_trait;
+use bytes::Bytes;
+use tokio::sync::mpsc::Sender;
+
+/// A trait for writing data to object store via multipart upload.
+/// The object won't be visible until the upload is completed.
+#[async_trait]
+pub trait MultipartWriter: Sync + Send {
+    /// Returns the part size in bytes. Most object stores require all parts excluding the last
+    /// one are the part size to be at least 5MB. And most object stores limit the max part number to be 10,000.
+    /// So the user can adjust the part size to upload an object more than 50GB.
+    fn part_size(&self) -> usize;
```

Comment:
This API is also a little bit wried because `MultipartBuffer` is a generic writer can be reused, but it requires a part size to split the input data, so we have to expose the part_size here.

Another approach is using `pub fn create_multipart_buffer(&self, tx: Sender<Bytes>) -> MultipartBuffer` instead.

Reply:
To be honest I don't know enough about other potential multipart writer implementations to give an informed comment on this. I am okay with this for now, we can always refactor to the other approach if we ever find that doing this will not work for a certain implementation.

Could you add a comment here with this information?

Reply:
Sounds good, thanks for the information. I'm happy with whatever you decide here! Let's just add a comment here with some information about the decision.